students/Roy_Tate/lesson04/mailroom2.py

```diff
@@ -0,0 +1,184 @@
+    print('Generating emails for all donors...')
+    count = 0
+    os.system('mkdir -p ' + save_path)
+    for donor, donations in donor_dict.items():
+        filename = save_path + donor + '.txt'.strip()
+        os.system('touch ' + filename)
+        with open(filename, 'w') as f:
+            f.write(email_str_annual.format(donor.title(), sum(donations)))
+            print('[SUCCESS] Output file -->  ' + str(os.path.abspath(filename)))
+            count += 1
+    print('[DONE] Successfully generated {} emails.'.format(str(count)))
+
+
+def send_thank_you():
+    # Obtain the name of the donor, the contribution amount, add the donation to the list, and generate
+    # a thank you email.
+
+    name = input("Enter the full name of the recipient (Type 'list' to view donors. Type '1' for Main Menu):\n")
+    # thank_you_args = {
+    #                'list': print_list,
+    #                '1': main,
+    # }
+    # thank_you_prompt = "Enter the full name of the recipient (Type 'list' to view donors. " \
+    #                    "Type '1' for Main Menu):\n"
+
+    # menu_selection(thank_you_prompt, thank_you_args)
+    if name == 'list':
+        print_list()
+
+    elif name == '1':
+        main()
+
+    elif name in donor_dict.keys():
+        donation_amount = float(input('EXISTING: Enter the donation amount: '))
+        add_donation(name, donation_amount)
+        print_email(email_str_ty, name, donation_amount)
+    else:  # name not found in list
+        donation_amount = float(input('NEW: Enter the donation amount: '))
+        add_donation(name, donation_amount)
+        print_email(email_str_ty, name, donation_amount)
+
+    # print(email_str.format(name.title(), donation_amount))
+
+
+# helper function
+def sum_donations(donor_dict):
+    return donor_dict[1]
+
+
+def create_report():
+    # generate a report of the current donor list, the total donation amounts, avg donation amt, and # of donations
+    header = '{:<20}|{:^15}|{:^10}|{:>15}'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+    header_len = len(header)
+    dotted_line = '\n{:<20}'.format('-' * header_len)
+    print(header + dotted_line)
+    donor_str_fmt = '{:<20} ${:>13.2f}{:>11}    ${:>11.2f}'
+    all_str = ''
+    sort_list = []
+    for name, donations in donor_dict.items():
+        num_gifts = len(Counter(donations))
+        total_given = sum(donations)
+        avg_gift = total_given / num_gifts
+        sort_list.append([name, total_given, num_gifts, avg_gift])
+    sort_list.sort(key=sum_donations, reverse=True)
+    for donors in sort_list:
+        all_str += donor_str_fmt.format(donors[0], donors[1], donors[2], donors[3]) + '\n'
+    print(all_str)
+
+
+# take a prompt as a string and a list of options/arguments as a dict
+def menu_selection(prompt, arg_dict):
+    while True:
+        try:
+            response = input(prompt)
+            arg_dict[response]() == ''
+        except:
```

Comment:
You've got the right idea here with the try/except block and we'll get into exceptions in more detail in lesson 5. However, you always want to catch a specific exception rather use a bare except statement like this. The problem with a bare except is that it will catch any exception that's raised, rather than just one (or more) that you expect to occur, and this can lead to bugs when errors pass through your code 'silently'. 

There's currently a bug like this in your program where if I try to exit from the command line by entering '4' to select the quit option, I see the following message and return to the original menu selection prompt:
```
Exiting the mailroom2.
Invalid Response. Try again.
```
What's happening here is that when your `die()` function calls `sys.exit()`, that function tries to exit the interpreter by raising the `SystemExit()` exception. Exceptions in Python will 'bubble up' the function call stack if they're not handled, so when this exception reaches your `except` statement in the `menu_selection()` function, it is caught and 'handled' by printing the error message and returning to the main prompt in your while loop.

It seems like the exception that you'd want to catch here is `KeyError`, because that will be raised if the user enters a value that is not a key in your `arg_dict`. You also want to try and only include code in your `try` block that might actually raise the exception, so I would suggest moving the call to input out of your try block and just including the call to the `arg_dict` function in there. Check out the course materials on exception handling in Lesson 5 for more details on best practices, you'll be required to use exception handling in mailroom for that assignment so you're getting a head start.